src/java.base/share/classes/sun/nio/cs/StreamEncoder.java

```diff
@@ -430,6 +430,10 @@ void implClose() throws IOException {
 
             if (bb.position() > 0)
                 writeBytes();
+        } catch (IOException x) {
+            encoder.reset();
+            throw x;
+        } finally {
             if (ch != null)
                 ch.close();
             else {
```

Comment:
If this encoder is for an OutputStreamWriter, and the flush of the left over characters fails, then the finally block will attempt to flush the underlying output stream again. Need to you think about that scenario a bit to decide if the underlying flush should be done for this case or not.

Reply:
> Need to think about that scenario a bit to decide if the underlying flush should be done for this case or not.

`flushLeftoverChar()` can throw an exception after having invoked `writeBytes()` once or more. This suggests that a flush of the underlying stream might be warranted.